### PR TITLE
Introducing CompilerContext.Module

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Avoids references to `runtime.Module` from `Compiler` & co.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [x] Existing unit tests continue to pass.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
Tests seem to be green and [I've fixed](https://github.com/enso-org/enso/pull/8144/commits/1338dcc18df7818b00e913a2b93dbea6c5db1dd3) the formatting. Let's review the idea and integrate if it is acceptable.

File: engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import java.util.Objects;
 
 import org.enso.compiler.Compiler;
+import org.enso.compiler.context.CompilerContext;
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java]: spurious?

File: engine/runtime/src/main/java/org/enso/compiler/context/CompilerContext.java [hunk 2]
```diff
@@ -47,8 +51,11 @@ public interface CompilerContext {
 
   void notifySerializeModule(QualifiedName moduleName);
 
+  // XXX delete
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/context/CompilerContext.java]: ?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java [hunk 3]
```diff
@@ -519,6 +530,11 @@ void setHasCrossModuleLinks(boolean hasCrossModuleLinks) {
     this.hasCrossModuleLinks = hasCrossModuleLinks;
   }
 
+  // XXX
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java]: spurious?

File: engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala [hunk 4]
```diff
@@ -476,11 +486,17 @@ private class DefaultPackageRepository(
...
+      loadedModules.put(
+        syntheticModule.getName.toString,
+        syntheticModule.asCompilerModule()
+      )
     } else {
       val loaded = loadedModules(syntheticModule.getName.toString)
       assert(!loaded.isSynthetic)
-      loaded.setDirectModulesRefs(refs.asJava)
+      loaded
+        .asInstanceOf[TruffleCompilerContext.Module]
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala:497]: this casting cannot be hidden as well?

File: engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala [hunk 5]
```diff
@@ -531,7 +547,10 @@ private class DefaultPackageRepository(
       key    <- keys
       module <- loadedModules.remove(key)
     } {
-      module.renameProject(newName)
+      module
+        .asInstanceOf[TruffleCompilerContext.Module]
+        .unsafeModule()
+        .renameProject(newName)
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala:553]: same comment as above

**Review by hubertp — APPROVED** [review]
Minor tweaks to be addressed

File: engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java]: Removed in b2aa71e9ca

File: engine/runtime/src/main/java/org/enso/compiler/context/CompilerContext.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/context/CompilerContext.java]: Deleted in 0e21cd4dbd

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java]: Documentation improved in f485719da8

File: engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala:497]: If we abstract away `PackageRepository` (we should as it references `TruffleFile`) and replace it with some methods in `CompilerContext`, then maybe. Maybe in later PR - removing this isn't essential for the startup speedup.

File: engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/interpreter/runtime/DefaultPackageRepository.scala:553]: Maybe - [as in the other reply](https://github.com/enso-org/enso/pull/8144#discussion_r1371802616).

**Pull request merged**